app/models/product.rb

```diff
@@ -57,11 +57,63 @@ def as_json_for_data
   #
   # `dist/` serialization
 
-  def as_json(options = {})
-    super(options.merge({ methods: :type }))
+  def as_json(integration_version: nil)
+    {
+      "attributes" => get_attributes,
+      "category" => get_category(integration_version:),
+    }.compact
   end
 
   def as_txt
     full_name.to_s
   end
+
+  private
+
+  def get_attributes
+    payload["properties"]&.map do |property|
+      {
+        "attribute" => Attribute.find_by(friendly_id: property["attribute"]).gid,
+        "value" => Value.find_by(friendly_id: property["value"])&.gid,
+      }
+    end
+  end
+
+  def get_category(integration_version:)
+    if payload["product_category_id"].is_a?(Array)
+      payload["product_category_id"].map do |category_id|
+        get_category_hash(category_id:, integration_version:)
+      end
+    else
+      category_id = parse_category_id(payload["product_category_id"])
+      get_category_hash(category_id:, integration_version:)
+    end
+  end
+
+  def parse_category_id(category_id)
+    category_id.split("/").last
+  end
+
+  def get_category_hash(category_id:, integration_version:)
+    category_id = parse_category_id(category_id)
+    if integration_version.nil?
+      Category.find_by(id: category_id)&.as_json&.slice("id", "full_name")
+    else
+      {
+        "id" => category_id,
+        "full_name" => integration_full_name(category_id: category_id, integration_version: integration_version),
+      }
+    end
+  end
+
+  def integration_full_name(category_id:, integration_version:)
+    full_name_map(integration_version:)[category_id.to_s]
+  end
+
+  def full_name_map(integration_version:)
+    categories = YAML.load_file(File.join(Rails.root, "data/integrations/#{integration_version}/full_names.yml"))
```

Comment:
It's better to have the YAML loaded once, instead of each iteration (It happened to me on the pipeline and the difference on the time of processing was huge)

Reply:
Good catch! I've memoized it on a per integration basis now at the class level, so it should be a little more speedy.